File: src/omnia/data/structure/observable/writable/GenericObservableChannels.java
package omnia.data.structure.observable.writable;

import io.reactivex.rxjava3.core.Observable;
import omnia.data.structure.observable.ObservableDataStructure;

class GenericObservableChannels<
        StateType, MutationEventType extends ObservableDataStructure.MutationEvent>
    implements ObservableDataStructure.ObservableChannels {

  private final Observable<StateType> states;
  private final Observable<MutationEventType> mutations;

  protected GenericObservableChannels(
      Observable<StateType> states, Observable<MutationEventType> mutations) {
    this.states = states;
    this.mutations = mutations;
  }

  @Override
  public Observable<StateType> states() {
    return states;
  }

  @Override
  public Observable<MutationEventType> mutations() {
    return mutations;
  }
}

File: src/omnia/data/structure/observable/writable/GenericObservableChannels.kt
package omnia.data.structure.observable.writable

import io.reactivex.rxjava3.core.Observable
import omnia.data.structure.observable.ObservableDataStructure

internal open class GenericObservableChannels<StateType, MutationEventType : ObservableDataStructure.MutationEvent> protected constructor(
        private val states: Observable<StateType>, private val mutations: Observable<MutationEventType>) : ObservableDataStructure.ObservableChannels {
    override fun states(): Observable<StateType> {
        return states
    }

    override fun mutations(): Observable<MutationEventType> {
        return mutations
    }
}